k8s enhancements

yorinasub17 opened the pull request:
More functions for https://github.com/gruntwork-io/terratest/issues/194

- Expose some functions that can be used without `testing.T` to not take in `testing.T` so that they can be used outside testing context.
    - `GetPublicIpsOfEc2InstancesFromClientE`
    - `GetKubernetesClientFromFileE`
    - `GetNodesByClientE`

- Migrate `ContinuouslyCheckUrl` from module-ecs
- Expose `StoreConfigToFile` from the k8s module tests, as that can be used for storing a kubernetes resource config to disk to be used with `kubectl apply`
- Introduce wrapper functions for calling kubectl:
     - `RunKubectl`, `RunKubectlE`, `RunKubectlAndGetOutputE`
     - `KubectlApplyFromString`, `KubectlApplyFromStringE`
     - `KubectlDeleteFromString`, `KubectlDeleteFromStringE`

- Introduce functions for interacting with Kubernetes Service resources.

brikis98 commented:
> Expose some functions that can be used without testing.T to not take in testing.T so that they can be used outside testing context.

What's the use case? Do we really want production code depending on a test library?

> Migrate ContinuouslyCheckUrl from module-ecs

Nice 👍

brikis98 commented on modules/k8s/config.go:
```diff
@@ -23,6 +24,19 @@ func LoadConfigFromPath(path string) clientcmd.ClientConfig {
 
+// LoadApiClientConfig will load a ClientConfig object from a file path that points to a location on disk containing a
+// kubectl config, with the requested context loaded.
+func LoadApiClientConfig(path string, context string) (*restclient.Config, error) {
```
`LoadApiClientConfigE`

yorinasub17 replied:
Fixed

brikis98 commented on modules/k8s/kubectl.go:
```diff
@@ -0,0 +1,101 @@
+// filename.
+func StoreConfigToTempFile(t *testing.T, configData string) string {
+	out, err := StoreConfigToTempFileE(t, configData)
+	require.NoError(t, err)
```
That is cleaner. Will make mental note to use this elsewhere in Terratest in the future.

brikis98 commented on modules/k8s/service.go:
```diff
@@ -0,0 +1,211 @@
+	clientset, err := GetKubernetesClientE(t)
+	require.NoError(t, err)
+	statusMsg := fmt.Sprintf("Wait for service %s to be provisioned.", serviceName)
+	message, err := retry.DoWithRetryE(
```
Could use the non "E" version here

yorinasub17 replied:
Fixed

brikis98 commented on modules/k8s/service.go:
```diff
@@ -0,0 +1,211 @@
+		// The load balancer is ready if it has at least one ingress point
+		return len(ingress) > 0
+	default:
+		return true
```
Shouldn't the default be `false`?

yorinasub17 replied:
See comment at beginning of function:
> Only the LoadBalancer type has a delay. All other service types are available if the resource exists.

brikis98 commented on modules/k8s/service.go:
```diff
@@ -0,0 +1,211 @@
+		return corev1.Node{}, err
+	}
+	if len(nodes) == 0 {
+		return corev1.Node{}, NewKubernetesError("There are no nodes in the cluster")
```
What's the advantage of a single `NewKubernetesError` used throughout this code? It doesn't make it easy to type check for specific errors, unlike, for example, a `NoNodesInCluster` type that is specific to this exact error here.

yorinasub17 replied:
That's a good point. Will fix.

brikis98 commented on modules/k8s/service.go:
```diff
@@ -0,0 +1,211 @@
+	// Path is /AVAILABILITY_ZONE/INSTANCE_ID
+	parts := strings.Split(awsIDUri.Path, "/")
+	instanceID := parts[2]
+	availabilityZone := parts[1]
```
Check length of `parts`

yorinasub17 replied:
Added check

yorinasub17 commented:
> What's the use case? Do we really want production code depending on a test library?

While I agree this is a bit awkward, the wrappers we built around the clouds end up being useful abstractions for writing any code that works against the underlying systems.

For example, in the rollout command I ended up doing a lot of the same things I would do in a test context: waiting for nodes to be ready, initializing clients by config files, manipulating config files, calling `kubectl`. And I really wanted to avoid copy pasting the code and duplicating the tests.

I guess the most ideal situation would be to break out the cloud verification code into its own repo that then prod code and terratest can import, but I wasn't quite ready to shave the yak of figuring out the right abstraction, and also cause a lot of disruption.

yorinasub17 commented on modules/http-helper/continuous.go:
```diff
@@ -0,0 +1,29 @@
+				statusCode, body, err := HttpGetE(t, url)
+				logger.Logf(t, "Got response %d and err %v from URL at %s", statusCode, err, url)
+				if err != nil {
+					t.Fatalf("Failed to make HTTP request to the URL at %s: %s\n", url, err.Error())
```
Yes this works. It's been causing my tests to fail (and me pulling hair) the past 2 business days 😄 Also, I copy pasted this with no modifications out of `module-ecs`.

brikis98 replied:
Hmm...

https://github.com/ipfs/go-ipfs/issues/2043
https://github.com/golang/go/issues/24678
https://github.com/golang/go/issues/15758

yorinasub17 replied:
Ah ok I see. So the bug is that there are race conditions in there. Ok I'll see what I can do to fix this.

brikis98 replied:
Ah, nice find! You should add a comment that indicates why we specifically use `Errorf` to each spot in the code so we don't accidentally break that.

One other way to do this would be for the method to return a channel and to write to that channel an error if something went wrong or `nil` if everything finished successfully. The calling code could then read from that channel to see how things went. But `Errorf` is probably good enough for now that it's not worth adding this complexity.

brikis98 commented:
> For example, in the rollout command I ended up doing a lot of the same things I would do in a test context: waiting for nodes to be ready, initializing clients by config files, manipulating config files, calling kubectl. And I really wanted to avoid copy pasting the code and duplicating the tests.

Sanity check: is this code _really_ the same or just _somewhat_ the same. I worry that Terratest may update a lot of these functions over time to be more and more suited for testing, with little regard for potential prod uses of this code... And if it's not too much code, sometimes, trying to keep things perfectly DRY can be counter-productive.

yorinasub17 commented:
> is this code really the same or just somewhat the same.

The kubectl wrappers (`RunKubectl`) are exactly the same right now. Getting and waiting for nodes are also exactly the same. The AWS stuff is more questionable.

> I worry that Terratest may update a lot of these functions over time to be more and more suited for testing, with little regard for potential prod uses of this code...

Those are valid points, although as mentioned, it would be nice if they were isolated in a separate library to avoid that... but then there is overhead ⬇️ 

> And if it's not too much code, sometimes, trying to keep things perfectly DRY can be counter-productive.


Ok I will revert those and copy paste and see how much duplication it ends up being, and if overtime it really does get out of hand, we can discuss merging then.

yorinasub17 commented:
- Reverted the "FromClient" versions
- Addressed all suggested changes:
    - Added more unique error types
    - Use the `E` suffix convention

yorinasub17 commented:
UPDATES:
- c7853c2 - Made `ContinuouslyCheckUrl` use `Errorf`. It will also now return a `sync.WaitGroup` that can be used to wait for the continuous check to shutdown.
- ad2c714 - Found and fixed a bug in `DummyServer`, where we try to register multiple handlers to route `/` on the `DefaultServeMux`. Caught this when I wrote a test for `ContinuouslyCheckUrl`.
- 4f285bf - Added more k8s functionalities, related to dealing with namespaces. Also changed the function APIs to take in a `KubectlOptions` object so that you can control which context to use as well as which namespace.
- 8d4160d - Added a basic k8s example and test

Currently, I am waiting to get a green build.

yorinasub17 commented:
UPDATE:
- 0091cc6 - Introduce `GetRandomStableRegion` for `aws`, which only selects from regions that have been around for more than a year. The list is hardcoded for now, but added the reference so that it can be evaluated on a yearly basis. This way, we won't suddenly have a bunch of test failures everytime a new region pops up.

brikis98 commented on examples/kubernetes-basic-example/README.md:
```diff
@@ -0,0 +1,40 @@
+# Kubernetes Basic Example
```
Nice 👍

brikis98 replied:
BTW, you may want to add to our newsletter that Terratest now supports automated testing for Kubernetes as a top-level item. SEO gold.

yorinasub17 replied:
Yup was thinking of doing that as soon as this merges

brikis98 commented on modules/aws/region.go:
```diff
@@ -20,6 +20,35 @@ const regionOverrideEnvVarName = "TERRATEST_REGION"
 // this region as a default.
 const defaultRegion = "us-east-1"
 
+// Reference for launch dates: https://aws.amazon.com/about-aws/global-infrastructure/
```
Nice addition 👍

brikis98 replied:
BTW, is there any easy way to add a yearly reminder to update this? Something in GitHub? Calendar invite?

yorinasub17 replied:
Set a reminder in slack for now. Used 1/23 every year.

brikis98 commented on modules/http-helper/dummy_server.go:
```diff
@@ -28,7 +28,9 @@ func RunDummyServer(t *testing.T, text string) (net.Listener, int) {
-	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+	// Create new serve mux so that multiple handlers can be created
+	server := http.NewServeMux()
+	server.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
```
Nice fix. I could've sworn we did this already at some point... But it must've been in some other code base...

brikis98 commented on modules/http-helper/dummy_server_test.go:
```diff
@@ -22,6 +23,25 @@ func TestRunDummyServer(t *testing.T) {
+		stopChecking <- true
+		wg.Wait()
+		shutDownServer(t, listener)
+	}()
```
In theory, this test could pass if `ContinuouslyCheckUrl` was a noop... Worth adding a negative test that takes the server down part of the way through and checks that a failure happened? This might be tricky to do given that `ContinuouslyCheckUrl` calls `t.Errorf` explicitly... Unless you pass it a mock `t`...

yorinasub17 replied:
Ok what do you think of? 4b54663 Not perfect, but I think is slightly better.

I know we can do the error channel instead like you suggested above, but I like the fact that the test will fail without any additional logic if there is one request that failed in the current model.

brikis98 replied:
LGTM!

brikis98 commented on modules/k8s/errors.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// ServiceNotAvailable is returned when a Kubernetes service is not yet available to accept traffic.
+type ServiceNotAvailable struct {
```
These are much clearer errors, thanks 👍

brikis98 commented on test/kubernetes_basic_example_test.go:
```diff
@@ -0,0 +1,47 @@
+	// This will get the service resource and verify that it exists and was retrieved successfully. This function will
+	// fail the test if the there is an error retrieving the service resource from Kubernetes.
+	service := k8s.GetService(t, options, "nginx-service")
+	require.Equal(t, service.Name, "nginx-service")
```
Does this test actually try to connect to nginx?

yorinasub17 replied:
This one, no. I wanted to keep this example as basic as possible. There is another test that will actually connect and check nginx: `kubernetes_basic_example_service_check_test.go`. You can see there that there is some added complexity from getting the endpoint and verifying it.

yorinasub17 commented:
Ok build passed, and review checked off. Going to merge, release, and start drafting an entry in the newsletter!
